src/BCReader.jl

```diff
@@ -0,0 +1,310 @@
+- `comms_ctx`: [ClimaComms.AbstractCommsContext] context used for this operation.
+- `interpolate_daily`: [Bool] switch to trigger daily interpolation.
+- `segment_idx0`: [Vector{Int}] index of the file data that is closest to date0.
+- `scaling function`: [Function] scales, offsets or transforms `varname`.
+- `land_mask`: [Fields.field] mask with 1 = land, 0 = ocean / sea-ice.
+- `date0`: [Dates.DateTime] start date of the file data.
+- `mono`: [Bool] flag for monotone remapping of `datafile_rll`.
+
+# Returns
+- BCFileInfo
+"""
+function bcfile_info_init(
+    FT,
+    bcfile_dir,
+    datafile_rll,
+    varname,
+    boundary_space;
+    comms_ctx = ClimaComms.SingletonCommsContext(),
+    interpolate_daily = false,
+    segment_idx0 = nothing,
+    scaling_function = no_scaling,
+    land_mask = nothing,
+    date0 = nothing,
+    mono = true,
+)
+
+    # regrid all times and save to hdf5 files
+    hd_outfile_root = varname * "_cgll"
+    if ClimaComms.iamroot(comms_ctx)
+        Regridder.hdwrite_regridfile_rll_to_cgll(
+            FT,
+            bcfile_dir,
+            datafile_rll,
+            varname,
+            boundary_space;
+            hd_outfile_root = hd_outfile_root,
+            mono = mono,
+        )
+    end
+    ClimaComms.barrier(comms_ctx)
+    data_dates = JLD2.load(joinpath(bcfile_dir, hd_outfile_root * "_times.jld2"), "times")
+
+    # init time tracking info
+    current_fields = Fields.zeros(FT, boundary_space), Fields.zeros(FT, boundary_space)
+    segment_length = [Int(0)]
+
+    # unless the start file date is specified, find the closest one to the start date
+    segment_idx0 =
+        segment_idx0 != nothing ? segment_idx0 :
+        [
+            argmin(
+                abs.(
+                    parse(FT, CallbackManager.datetime_to_strdate(date0)) .-
+                    parse.(FT, CallbackManager.datetime_to_strdate.(data_dates[:]))
+                ),
+            ),
+        ]
+
+    return BCFileInfo(
+        FT,
+        bcfile_dir,
+        comms_ctx,
+        hd_outfile_root,
+        varname,
+        data_dates,
+        deepcopy(segment_idx0),
+        segment_idx0,
+        current_fields,
+        segment_length,
+        scaling_function,
+        interpolate_daily,
+        land_mask,
+    )
+end
+
+# IO - monthly
+"""
+    update_midmonth_data!(date, bcf_info)
+
+Extracts boundary condition data from regridded (to model grid) NetCDF files.
+The times for which data is extracted depends on the specifications in the
+`bcf_info` struct).
+
+# Arguments
+- `date`: [Dates.DateTime] start date for data.
+- `bcf_info`: [BCFileInfo] containing boundary condition data.
+"""
+function update_midmonth_data!(date, bcf_info::BCFileInfo)
+    # monthly count
+    all_dates = bcf_info.all_dates
+    midmonth_idx = bcf_info.segment_idx[1]
+    midmonth_idx0 = bcf_info.segment_idx0[1]
+    outfile_root = bcf_info.hd_outfile_root
+    scaling_function = bcf_info.scaling_function
```

Comment:
```suggestion
    (; hd_outfile_root, scaling_function) = bcf_info
```
Note that unpacking can be done like this, too